src/uct/tcp/tcp_cm.c

```diff
@@ -211,94 +204,146 @@ ucs_status_t uct_tcp_cm_send_event(uct_tcp_ep_t *ep,
+                            str, ucs_min(max_size, UCS_SOCKADDR_STRING_LEN));
+}
 
-    uct_tcp_iface_remove_ep(ep);
+static void
+uct_tcp_cm_conn_match_purge_cb(ucs_conn_match_ctx_t *conn_match_ctx,
+                               ucs_conn_match_elem_t *elem)
+{
+    uct_tcp_ep_t *ep = ucs_container_of(elem, uct_tcp_ep_t, elem);
 
-    ucs_list_add_tail(ep_list, &ep->list);
-    ucs_debug("tcp_iface %p: tcp_ep %p added to %p list",
-              iface, ep, ep_list);
+    /* EP was deleted from the connection matching context during cleanup
+     * procedure, move EP to the iface's EP list to correctly destroy EP */
+    ucs_assert(ep->flags & UCT_TCP_EP_FLAG_ON_MATCH_CTX);
+    ep->flags &= ~UCT_TCP_EP_FLAG_ON_MATCH_CTX;
+    uct_tcp_iface_add_ep(ep);
 
-    return UCS_OK;
+    uct_tcp_ep_destroy_internal(&ep->super.super);
 }
 
-void uct_tcp_cm_remove_ep(uct_tcp_iface_t *iface, uct_tcp_ep_t *ep)
-{
-    ucs_list_link_t *ep_list;
-    khiter_t iter;
+const ucs_conn_match_ops_t uct_tcp_cm_conn_match_ops = {
+    .get_address = uct_tcp_cm_conn_match_get_address,
+    .get_conn_sn = uct_tcp_cm_conn_match_get_conn_sn,
+    .address_str = uct_tcp_cm_conn_match_address_str,
+    .purge_cb    = uct_tcp_cm_conn_match_purge_cb
+};
 
-    iter = kh_get(uct_tcp_cm_eps, &iface->ep_cm_map, ep->peer_addr);
-    ucs_assertv(iter != kh_end(&iface->ep_cm_map), "iface=%p", iface);
+uct_tcp_cm_conn_sn_t
+uct_tcp_cm_get_conn_sn(uct_tcp_iface_t *iface,
+                       const struct sockaddr_in *dest_address)
+{
+    return (uct_tcp_cm_conn_sn_t)
+           ucs_conn_match_get_next_sn(&iface->conn_match_ctx,
+                                      dest_address);
+}
 
-    ep_list = kh_value(&iface->ep_cm_map, iter);
-    ucs_assertv(!ucs_list_is_empty(ep_list), "iface=%p", iface);
+uct_tcp_ep_t *uct_tcp_cm_get_ep(uct_tcp_iface_t *iface,
+                                const struct sockaddr_in *dest_address,
+                                uct_tcp_cm_conn_sn_t conn_sn,
+                                uint8_t with_ctx_cap)
+{
+    ucs_conn_match_queue_type_t queue_type;
+    ucs_conn_match_elem_t *elem;
+    uct_tcp_ep_t *ep;
+    int remove_from_ctx;
+
+    ucs_assert(conn_sn < UCT_TCP_CM_CONN_SN_MAX);
+    ucs_assert((with_ctx_cap == UCT_TCP_EP_FLAG_CTX_TYPE_TX) ||
+               (with_ctx_cap == UCT_TCP_EP_FLAG_CTX_TYPE_RX));
+
+    /* Looking for EP with TX capability is a part of a connection request
+     * handling, the connection request could be received from a ghost
+     * connection that should be closed. So, need to search for the EP
+     * in any queue and don't remove EP in case of it was found. The ghost
+     * connection will removed as soon as the disconnection on a socket fd
+     * is detected bu EP. */
+    if (with_ctx_cap == UCT_TCP_EP_FLAG_CTX_TYPE_TX) {
+        queue_type      = UCS_CONN_MATCH_QUEUE_ANY;
+        remove_from_ctx = 0;
+    } else {
+        queue_type      = UCS_CONN_MATCH_QUEUE_UNEXP;
```

Comment:
/* when creating new endpoint from API, search for arrived connection requests */